File: workspace/solutions/go/s0/s3/s0/s4/s_0304_00/s_0304_00.go
/*
 * @lc app=leetcode.cn id=304 lang=golang
 *
 * [304] 二维区域和检索 - 矩阵不可变
 */

package s_0304_00

// @lc code=start
type TMatrix = [][]int

type NumMatrix struct {
	n                int
	m                int
	matrix           TMatrix
	prefixSumsMatrix TMatrix
	initialized      bool
}

func Constructor(matrix [][]int) NumMatrix {
	n := len(matrix)
	m := len(matrix[0])
	prefixSumsMatrix := make([][]int, n+1)
	for i := range prefixSumsMatrix {
		prefixSumsMatrix[i] = make([]int, m+1)
	}

	return NumMatrix{
		n:                n,
		m:                m,
		matrix:           matrix,
		prefixSumsMatrix: prefixSumsMatrix,
		initialized:      false,
	}
}

func (this *NumMatrix) SumRegion(row1 int, col1 int, row2 int, col2 int) int {
	this.init()

	return this.prefixSumsMatrix[row2+1][col2+1] -
		this.prefixSumsMatrix[row2+1][col1] -
		this.prefixSumsMatrix[row1][col2+1] +
		this.prefixSumsMatrix[row1][col1]
}

func (this *NumMatrix) init() bool {
	if this.initialized {
		return false
	}

	for i := 0; i < this.n; i++ {
		for j := 0; j < this.m; j++ {
			this.prefixSumsMatrix[i+1][j+1] = this.prefixSumsMatrix[i+1][j] +
				this.prefixSumsMatrix[i][j+1] -
				this.prefixSumsMatrix[i][j] +
				this.matrix[i][j]
		}
	}

	this.initialized = true
	return true
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * obj := Constructor(matrix);
 * param_1 := obj.SumRegion(row1,col1,row2,col2);
 */
// @lc code=end

func (this NumMatrix) Func(row1 int, col1 int, row2 int, col2 int) int {
	return this.SumRegion(row1, col1, row2, col2)
}

var Solution = Constructor

File: workspace/solutions/java/lib/src/main/java/s0/s3/s0/s4/s_0304_00/s_0304_00.java
/*
 * @lc app=leetcode.cn id=304 lang=java
 *
 * [304] 二维区域和检索 - 矩阵不可变
 */

package s0.s3.s0.s4.s_0304_00;

// @lc code=start
class NumMatrix {
    private final int n;
    private final int m;
    private final int[][] matrix;
    private final int[][] prefix_sums_matrix;
    private boolean initialized = false;

    public NumMatrix(
        int[][] matrix
    ) {
        this.n = matrix.length;
        this.m = matrix[0].length;
        this.matrix = matrix;
        this.prefix_sums_matrix = new int[this.n + 1][this.m + 1];
    }

    public int sumRegion(
        int row1, int col1, int row2, int col2
    ) {
        this.init();
        return this.prefix_sums_matrix[row2 + 1][col2 + 1]
            - this.prefix_sums_matrix[row2 + 1][col1]
            - this.prefix_sums_matrix[row1][col2 + 1]
            + this.prefix_sums_matrix[row1][col1];
    }

    private boolean init() {
        if (this.initialized) {
            return false;
        }

        for (int i = 0; i < this.n; i++) {
            for (int j = 0; j < this.m; j++) {
                this.prefix_sums_matrix[i + 1][j + 1] //
                    = this.prefix_sums_matrix[i + 1][j]
                    + this.prefix_sums_matrix[i][j + 1]
                    - this.prefix_sums_matrix[i][j]
                    + this.matrix[i][j];
            }
        }

        this.initialized = true;
        return true;
    }
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix obj = new NumMatrix(matrix);
 * int param_1 = obj.sumRegion(row1,col1,row2,col2);
 */
// @lc code=end

public class s_0304_00 extends NumMatrix {
    public s_0304_00(
        int[][] matrix
    ) {
        super(matrix);
    }
}
